### ISSUE TITLE
Force casting mask to np.uint16

### ISSUE THREAD

**Issue by yfukai** [MouseLand/cellpose#283]
Hi, thank you for this great program! 

I'm trying to segment a large image, and noticed that a mask with the index >65535 is not correctly indexed (there appears several regions with the same index). I think this is because the mask is cast to `np.uint16` around this line, though it is treated as `np.int32` beforehand.
https://github.com/MouseLand/cellpose/blob/babf2eff59265815f0462f8cbc9fb9c68e747cc3/cellpose/models.py#L593
Does it make sense to change `np.uint16` to `np.int32` or `np.uint32`?

Cheers,
YF

**Comment by yfukai** [comment]
Opened a pull request in case it's appropriate, but any suggestions are welcome!

**Comment by kevinjohncutler** [comment]
@yfukai good catch, I recently ran into this bug myself for some of the ground-truth labels that used int32 labels and -1 for background. I wrote a function to format the labels properly for training that may also be useful here. Does your image actually have more than 2^16-1 cells? Since I don't think the signed->unsigned type change is the culprit for masks generated _by_ cellpose, you must be right that it must be a cell count issue... but that number of cells seems crazy! It could also just be extreme oversegmentation. Could you share what the input and output look like?

**Comment by yfukai** [comment]
Hi, @kevinjohncutler , thanks for your comment! 
I'm actually trying to segment a fused 11 x 11 tile of images, approximately 14000 x 14000 in pixel. (which is huge but a kind of realistic scale I believe!) The cell diameter is, say, 15px , so even considering the cell density of 1/5, > 65536 is not that strange number. Actually, by visual check, the segmentation itself looks OK.
![image](https://user-images.githubusercontent.com/5919272/128300960-937b95ad-b770-4629-828d-a072038fbe95.png)

**Comment by yfukai** [comment]
A temporal solution for those who would like the feature in [the pull request](https://github.com/MouseLand/cellpose/pull/284) for large-image segmentation: 
```bash
pip uninstall cellpose
pip install git+https://github.com/yfukai/cellpose.git@fix_mask_dtype
```

**Comment by elmbeech** [comment]
@yfukai for this timely bug fix!
Interestingly, we are running cell pose for a while now, and I am sure older (mxnet based?) version could handle >= 2**16 cells.
Just to confirm, we usually segment > 20000px *  20000px scenes and it is not uncommon to detect more then that number  of nuclei in tissue sample.

**Comment by yfukai** [comment]
Hi @elmbeech, thank you for your comment!
I was reading the [old code](https://github.com/MouseLand/cellpose/blob/5402a3307e59f2035a18246701dab87bb6017979/cellpose/dynamics.py#L410) and as far as understand there was no cast to `np.uint16` in the `mxnet` version. I guess that's why it works for large image in your case.

**Comment by carsen-stringer** [comment]
this should be working now

**Comment by FiReTiTi** [comment]
Is this issue fixed?
I'm running CellPose 6.5 and I have the same issue for a big image containing more than 65535 nuclei.

**Comment by elmbeech** [comment]
cellpose 6.5?
the latest version is 2.0.
and yes, it is fixed. it worked for us.

Elmar
________________________________________
From: FiReTiTi ***@***.***>
Sent: Monday, April 25, 2022 12:46:21 PM
To: MouseLand/cellpose
Cc: Elmar Bucher; Mention
Subject: [EXTERNAL] Re: [MouseLand/cellpose] Force casting mask to np.uint16 (#283)

Is this issue fixed?
I'm running CellPose 6.5 and I have the same issue for a big image containing more than 65535 nuclei.

**Comment by FiReTiTi** [comment]
My bad, it's obviously 0.6.5.
@elmbeech In which version was it fixed?

**Comment by elmbeech** [comment]
if I remember right, version 1.0.
just update to the latest version, and it is fixed for sure.

best, Elmar
________________________________________
From: FiReTiTi ***@***.***>
Sent: Monday, April 25, 2022 3:42:55 PM
To: MouseLand/cellpose
Cc: Elmar Bucher; Mention
Subject: [EXTERNAL] Re: [MouseLand/cellpose] Force casting mask to np.uint16 (#283)

My bad, it's obviously 0.6.5.
@elmbeech<https://urldefense.com/v3/__https://github.com/elmbeech__;!!Mi0JBg!M6ZNKsVEqeLY9RgQaSmISXG6H4WA2AIjy78vYhjCWUzbguAMxXbPI6hPWPnmUqRHzDGwjV0SssySR1YNOSmpMv9w$> In which version was it fixed?

**Comment by FiReTiTi** [comment]
Fixed on version 1.0.2.
Thanks.
